File: 0077-combinations/0077-combinations.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        def combination(index):
            if len(route) == k:
                result.append(list(route))
                return

            for i in range(index, n + 1):
                if not visited[i]:
                    route.add(i)
                    visited[i] = True
                    combination(i + 1)
                    route.remove(i)
                    visited[i] = False

        visited = [False for i in range(n + 1)]
        result = list()
        route = set()

        combination(1)

        return result

File: 0077-combinations/0077-combinations.java
class Solution {
    private List<List<Integer>> result;
    private boolean[] visited;
    private Set<Integer> route;
    private int n;
    private int k;

    private void combination(int index) {
        if (route.size() == k) {
            result.add(List.copyOf(route));
            return;
        }

        for (int i = index; i < n + 1; i++) {
            if (!visited[i]) {
                route.add(i);
                visited[i] = true;
                combination(i + 1);
                route.remove(i);
                visited[i] = false;
            }
        }
    }

    public List<List<Integer>> combine(int n, int k) {
        result = new ArrayList<>();
        visited = new boolean[n + 1];
        route = new HashSet<>();
        this.n = n;
        this.k = k;

        combination(1);

        return result;
    }
}